text/0000-cargo-rustup-discovery.md

```diff
@@ -0,0 +1,457 @@
+In the meantime, the error message will suggest setting `CARGO_SAFE_DIRECTORIES` or editing a config file.
+
+## Cargo config search enhancements
+
+There have been longstanding issues with how Cargo searches for config files.
+This proposal makes a change that may complicate future changes to that discovery process.
+Unfortunately, a plan to change the config discovery process is not on the horizon,
+so it is unknown what impact this will have.
+
+## Rustup `rust-toolchain` trust
+
+There have been discussions about adding a mechanism to Rustup to add some kind of "trust" system for `rust-toolchain` files.
+This idea would have an even stronger indication of the user being happy with a particular `rust-toolchain` file, since its contents would be hashed, and the trust would be revoked if it changes.
+
+It is not clear to me how this explicit trust system would interact with "safe directories".
+If it is added in the future, there may be some overlap or confusion on the difference between the two.
+
+## Granular capabilities
+
+An alternative is to associate actions and config options with granular capabilities.
+Then, Cargo and Rustup could trust files that are owned by different users if they don't
+present a hazard (such as executing external commands).
+This is the approach being taken by [gitoxide](https://github.com/Byron/gitoxide).
+It has a [`git-sec`](https://github.com/Byron/gitoxide/discussions/425) crate defining a security model.
+
+The Git developers also considered only generating an error if the ownership differed from the current user *and* there is a `core.fsmonitor` setting in the repo.
+However, they decided not to go that route.
+
+We feel like it is unlikely we can securely add such a mechanism to Cargo and Rustup due to the complexity.
+Also, in many cases, Cargo projects run code on the host
+(such as proc-macros and build scripts)
+which would severely limit what would be constituted as "safe".
+
+## Ceiling directories
+
+Git also has `GIT_CEILING_DIRECTORIES` for limiting the upwards search,
+and was presented as a way for git users to protect themselves.
+This can be used to avoid potentially loading malicious repositories,
+but in practice it is best used for improving performance.
+It might be nice for Cargo and Rustup to adopt something similar,
+but it is not a replacement for ownership checks since it is opt-in.
+
+## Honor the user of the current directory
+
+The ownership check could also include the owner of the current working directory as "approved" and allow any files owned by that user in any parent directory.
+This is under the assumption that the user running the command makes an explicit choice as to which directory they run Cargo or Rustup from.
+This could significantly ease the burden of dealing with this issue,
+as it will avoid needing to set safe directories in almost all use cases where they would be needed.
+
+However, I feel like this approach is too risky.
```

Comment:
OTOH this makes use-cases that use group-wide permissions (i.e. `rwxrwx---`) and a umask setup in a certain way work out of the box.

I'm somewhat invested in this, since I tried to utilize that sort of approach, though not with Rust...